### PR TITLE
Add hardware debug ROB

### PR DESCRIPTION
**Pull request by joey0320**
**Related issue**: 

**Type of change**: bug report | feature request | other enhancement

**Impact**: no functional change | API addition (no impact on existing code) | API modification

**Development Phase**: proposal |  implementation

**Release Notes**

### CODE REVIEW

File: src/main/scala/tile/FPU.scala [hunk 1]
```diff
@@ -727,7 +727,9 @@ class FPU(cfg: FPUParams)(implicit p: Parameters) extends FPUModule()(p) {
   val io = IO(new FPUIO)
 
   val (useClockGating, useDebugROB) = coreParams match {
-    case r: RocketCoreParams => (r.clockGate, r.debugROB)
+    case r: RocketCoreParams =>
+      val sz = if (r.debugROB.isDefined) r.debugROB.get.size else 1
+      (r.clockGate, sz < 1)
```
Comment (jerryz123) [inline src/main/scala/tile/FPU.scala:732]: Hmmm, this means that there is no write tracking of writes into FPRs right?

Comment (joey0320) [reply src/main/scala/tile/FPU.scala:732]: Yep. Just integer registers for now.

Comment (jerryz123) [reply src/main/scala/tile/FPU.scala:732]: Ok then

File: src/main/scala/rocket/RocketCore.scala [hunk 2]
```diff
@@ -761,18 +761,37 @@ class Rocket(tile: RocketTile)(implicit p: Parameters) extends CoreModule()(p)
...
-      (wb_ctrl.wfd || (wb_ctrl.wxd && wb_waddr =/= 0.U)) && !csr.io.trace(0).exception,
-      wb_ctrl.wxd && wb_wen && !wb_set_sboard,
-      wb_waddr + Mux(wb_ctrl.wfd, 32.U, 0.U))
-
-    io.trace.insns(0) := DebugROB.popTrace(clock, reset, io.hartid)
-
-    DebugROB.pushWb(clock, reset, io.hartid, ll_wen, rf_waddr, rf_wdata)
+  if (rocketParams.debugROB.isDefined) {
+    val sz = rocketParams.debugROB.get.size
+    if (sz < 1) {
```
Comment (jerryz123) [inline src/main/scala/rocket/RocketCore.scala:766]:
```suggestion
    if (sz < 1) { // use unsynthesizable ROB
```

File: src/main/scala/rocket/RocketCore.scala [hunk 3]
```diff
@@ -761,18 +761,37 @@ class Rocket(tile: RocketTile)(implicit p: Parameters) extends CoreModule()(p)
...
+      DebugROB.pushTrace(clock, reset,
+        io.hartid, csr_trace_with_wdata,
+        (wb_ctrl.wfd || (wb_ctrl.wxd && wb_waddr =/= 0.U)) && !csr.io.trace(0).exception,
+        wb_ctrl.wxd && wb_wen && !wb_set_sboard,
+        wb_waddr + Mux(wb_ctrl.wfd, 32.U, 0.U))
+
+      io.trace.insns(0) := DebugROB.popTrace(clock, reset, io.hartid)
+
+      DebugROB.pushWb(clock, reset, io.hartid, ll_wen, rf_waddr, rf_wdata)
+    } else {
```
Comment (jerryz123) [inline src/main/scala/rocket/RocketCore.scala:778]:
```suggestion
    } else { // synthesizable ROB (no FPRs)
```

**Review by jerryz123 — CHANGES_REQUESTED** [review]

File: src/main/scala/subsystem/Configs.scala [hunk 4]
```diff
@@ -447,10 +447,10 @@ class WithFPUWithoutDivSqrt extends Config((site, here, up) => {
   }
 })
 
-class WithRocketDebugROB(enable: Boolean = true) extends Config((site, here, up) => {
+class WithRocketDebugROB(size: Int = 0) extends Config((site, here, up) => {
```
Comment (jerryz123) [inline src/main/scala/subsystem/Configs.scala:450]: This needs to support removing the DebugROB too. 
```WithRocketDebugROB(enable: Boolean = true, size: Int = 0)```

**Review by jerryz123 — CHANGES_REQUESTED** [review]
@joey0320 we want to be able to remove the debugrob where necessary, please add that support to the WithRocketDebugROB.

**Comment by jerryz123** [comment]
Please update the title of this PR so that the merge commit will convey more useful information. 
Then after @sequencer approves, we can merge

**Review by sequencer — APPROVED** [review]
LGTM, In the future, I'd like to migrate it to layer API for splitting RTL and verification codes

**Pull request merged**